src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -63,18 +63,32 @@ The trivial additive character (sending everything to `1`) is `(1 : add_char R R
 @[derive [comm_monoid, inhabited]]
 def add_char : Type (max u v) := (multiplicative R) →* R'
 
-instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, multiplicative R → R') :=
-monoid_hom.has_coe_to_fun
+end add_char_def
+
+section coe_to_fun
+
+variables {R : Type u} [add_monoid R] {R' : Type v} [comm_monoid R']
+
+/-- Interpret an additive character as a monoid homomorphism. -/
+def add_char.to_monoid_hom : (add_char R R') → (multiplicative R →* R') := id
+
+open multiplicative
+
+/-- Define coercion to a function so that it includes the move from `R` to `multiplicative R`.
+After we have proved the API lemmas below, we don't need to worry about writing `of_add a`
+when we want to apply an additive character. -/
+instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, R → R') :=
+{ coe := λ ψ x, ψ.to_monoid_hom (of_add x) }
+
+lemma coe_to_fun_apply (ψ : add_char R R') (a : R) : ψ a = ψ.to_monoid_hom (of_add a) := rfl
```

Comment:
If you don't want to enter the namespace upstairs, then certainly this needs to be prefixed. Otherwise it ends up in the root namespace.